go/types/simplify.go

```diff
@@ -236,3 +236,25 @@ func (tc *TypeCache) simplifyStructs(expectedName string, ts typeset, intersectS
 
 	return tc.makeStructType(expectedName, fields)
 }
+
+func replaceAndCollectStructTypes(tc *TypeCache, t *Type) (*Type, map[string]map[*Type]struct{}) {
+	collected := map[string]map[*Type]struct{}{}
```

Comment:
I decided to use *Type instead of hash.Hash here because I do not want to invoke writeValue at this level. I was thinking it would be fine to have duplicates at this level.